# Data Cleanup 

In [17]:
import numpy as np
import pandas as pd 

In [18]:
import warnings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
warnings.filterwarnings('ignore')

In [19]:
data = pd.read_csv("/kaggle/input/datasets/abhijeetsp015/ml-project-data/raw_dirty_data_saas.csv")

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117374 entries, 0 to 117373
Data columns (total 32 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   customer_id                  117374 non-null  object 
 1   company_size                 117374 non-null  object 
 2   industry                     117374 non-null  object 
 3   region                       117374 non-null  object 
 4   subscription_plan            117374 non-null  object 
 5   contract_type                117374 non-null  object 
 6   tenure_months                117374 non-null  object 
 7   monthly_revenue              112623 non-null  object 
 8   seats_licensed               112663 non-null  float64
 9   login_frequency              117374 non-null  float64
 10  feature_usage_score          111542 non-null  object 
 11  avg_session_duration_mins    117374 non-null  float64
 12  last_login_days_ago          110265 non-null  float64
 13 

In [21]:
data.head(10)

,customer_id,company_size,industry,region,subscription_plan,contract_type,tenure_months,monthly_revenue,seats_licensed,login_frequency,feature_usage_score,avg_session_duration_mins,last_login_days_ago,integration_count,support_tickets_count,avg_response_time,customer_satisfaction_score,payment_failures,discount_applied,onboarding_completed,email_open_rate,training_sessions_attended,device_type,signup_date,random_id_hash,theme_preference,ui_language,account_color,dummy_score,referral_code,last_password_change_days,churn
0,CUST-0028736,Mid-Market,Manufacturing,Europe,Basic,Monthly,2,193.61,16.0,14.3,57.9,23.7,NaN,3.0,4,3.8,5.0,2,0.0,No,0.726,3.0,Mixed,11 Sep 2024,52e1723aa5dc,Light,en-US,#EC4899,30.98,NaN,60,1
1,CUST-0104506,SMB,Technology,Europe,Basic,Yearly,26,210.59,7.0,9.6,55.5,27.8,44.0,4.0,2,6.6,4.4,0,50.0,Yes,0.288,3.0,Desktop,25 Aug 2022,da50e4213fca,Light,en-US,#3B82F6,7.39,REF-D066AD,16,0
2,CUST-0003425,Startup,Education,Europe,Basic,Yearly,62,147.22,6.0,20.6,97.0,46.5,8.0,4.0,4,19.9,5.0,0,10.0,Yes,NaN,4.0,Desktop,09-09-2019,4939174703a6,Light,de-DE,#EC4899,18.66,NaN,109,0
3,CUST-0025254,Startup,Education,North America,Enterprise,Monthly,22,1194.15,1.0,5.4,NaN,20.2,71.0,2.0,6,11.8,NaN,1,15.0,Yes,0.108,5.0,Desktop,08 Feb 2023,ed44d8a9fe3c,System Default,en-US,#3B82F6,69.04,REF-0F723E,63,0
4,CUST-0011252,Mid-Market,Education,APAC,Pro,Monthly,23,1104.54,19.0,12.4,78.1,32.3,NaN,4.0,2,7.2,NaN,0,0.0,No,0.855,1.0,Mobile,27/12/2022,0c833eae1082,Dark,en-US,#10B981,51.10,NaN,51,1
5,CUST-0064132,SMB,Logistics,North America,Basic,Monthly,10,147.91,4.0,10.4,76.5,35.7,28.0,4.0,1,3.5,5.0,0,15.0,Yes,0.491,NaN,Desktop,01-28-2024,418cfd9bb392,Dark,fr-FR,#3B82F6,12.45,NaN,120,0
6,CUST-0028883,SMB,Education,North America,Enterprise,Yearly,21,2499.54,4.0,14.0,55.9,39.8,66.0,3.0,3,1.3,5.0,0,0.0,Yes,0.595,1.0,Mixed,03-05-2023,51c99d313ae1,Light,de-DE,#EF4444,89.63,REF-73B23E,15,0
7,CUST-0089526,SMB,Technology,APAC,Basic,Monthly,60,206.12,7.0,4.9,16.5,15.9,82.0,1.0,6,10.6,4.3,0,25.0,No,0.750,2.0,Mobile,12-06-2019,d960ad94110b,Dark,zh-CN,#8B5CF6,0.96,REF-4AF55E,6,0
8,CUST-0049809,SMB,Manufacturing,North America,Basic,Yearly,43,163.92,8.0,14.3,62.3,30.6,59.0,3.0,3,1.5,5.0,2,0.0,Yes,0.532,3.0,Mobile,05-15-2021,d245f4d03038,Dark,en-US,#8B5CF6,10.88,NaN,40,0
9,CUST-0040975,SMB,Healthcare,Latin America,Enterprise,Monthly,5,2484.0,8.0,12.3,68.6,35.4,35.0,5.0,2,4.0,NaN,0,10.0,No,0.343,4.0,Desktop,22 May 2024,fc12016b721e,Dark,fr-FR,#3B82F6,60.07,NaN,110,0


### removing uncessary columns 

* random_id_hash — random hash, no ML value
* account_color — UI preference, not churn signal
* theme_preference — UI preference, not churn signal
* dummy_score — literally named dummy, 9,468 nulls


In [22]:
cols_to_drop = [
    'random_id_hash',
    'account_color',
    'theme_preference',
    'dummy_score',
    'referral_code'
]

data = data.drop(columns=cols_to_drop)
print("Remaining columns:", data.shape[1])

Remaining columns: 27


### Datatype change and handiling nulls in particular columns.
1. Tenure months to float
2. Monthly revenue to float
3. feature_usage_score to float
4. signup_date to date format mixed

data2['column_name'] = pd.to_numeric(data2['column_name'], errors='coerce') -- little unwanted values get formated to null using errors='coerce'. 
Lets clean them first. 

In [23]:
print('Null values before changing datatype')
print(data.isnull().sum()[['tenure_months','monthly_revenue','feature_usage_score','signup_date']])

Null values before changing datatype
tenure_months             0
monthly_revenue        4751
feature_usage_score    5832
signup_date            3521
dtype: int64


In [24]:
# Data Cleaning operations 
data['tenure_months'] = data['tenure_months'].str.strip().str.lower().str.replace("'", '')
data['monthly_revenue'] = data['monthly_revenue'].str.strip().str.lower().str.replace("'", '')
data['feature_usage_score'] = data['feature_usage_score'].str.strip().str.lower().str.replace("'", '')

In [25]:
# Changing Datatype 
data['tenure_months'] = pd.to_numeric(data['tenure_months'])
data['monthly_revenue'] = pd.to_numeric(data['monthly_revenue'])
data['feature_usage_score'] = pd.to_numeric(data['feature_usage_score'])
data['signup_date'] = pd.to_datetime(data['signup_date'], format="mixed")

In [26]:
print('Null values after changing datatype')
print(data.isnull().sum()[['tenure_months','monthly_revenue','feature_usage_score','signup_date']])

Null values after changing datatype
tenure_months             0
monthly_revenue        4751
feature_usage_score    5832
signup_date            3521
dtype: int64


In [27]:
print(data[['tenure_months','monthly_revenue','feature_usage_score','signup_date']].dtypes)

tenure_months                   int64
monthly_revenue               float64
feature_usage_score           float64
signup_date            datetime64[ns]
dtype: object


### Handling the duplicates
1. Initally from the main column - customer_id. Drop any duplicate customer to over train our model.


In [28]:
print("Duplicates before:", data.duplicated().sum())

data = data.drop_duplicates(subset='customer_id', keep='first') # removing on the basis of customer_id
data = data.reset_index(drop=True) # resetting the index 

Duplicates before: 1734


In [29]:
print("Duplicates after:", data.duplicated().sum())
print("New shape:", data.shape)

Duplicates after: 0
New shape: (115640, 27)


### Data Inconsistency

In [30]:
non_numeric_cols = data.select_dtypes(exclude='number').columns

for col in non_numeric_cols:
    print(f"Column: {col}")
    print("Unique Values:", data[col].nunique())
    print(data[col].dropna().unique()[:20])
    print("-" * 50)

Column: customer_id
Unique Values: 115640
['CUST-0028736' 'CUST-0104506' 'CUST-0003425' 'CUST-0025254'
 'CUST-0011252' 'CUST-0064132' 'CUST-0028883' 'CUST-0089526'
 'CUST-0049809' 'CUST-0040975' 'CUST-0025021' 'CUST-0089485'
 'CUST-0039694' 'CUST-0060873' 'CUST-0005671' 'CUST-0082085'
 'CUST-0046298' 'CUST-0024590' 'CUST-0012553' 'CUST-0088411']
--------------------------------------------------
Column: company_size
Unique Values: 4
['Mid-Market' 'SMB' 'Startup' 'Enterprise']
--------------------------------------------------
Column: industry
Unique Values: 10
['Manufacturing' 'Technology' 'Education' 'Logistics' 'Healthcare'
 'Marketing' 'Retail' 'Legal' 'Real Estate' 'Finance']
--------------------------------------------------
Column: region
Unique Values: 5
['Europe' 'North America' 'APAC' 'Latin America' 'Middle East & Africa']
--------------------------------------------------
Column: subscription_plan
Unique Values: 19
['Basic' 'Enterprise' 'Pro' 'Free' 'enterprise' 'PRO' 'Pro '

In [31]:
print(data['subscription_plan'].unique())
print(data['contract_type'].unique())
print(data['onboarding_completed'].unique())

['Basic' 'Enterprise' 'Pro' 'Free' 'enterprise' 'PRO' 'Pro ' 'Free '
 'BASIC' 'ENTERPRISE' 'free' 'pro ' 'BASIC ' 'Basic ' 'pro' 'FREE' 'basic'
 'Enterprise ' 'Ent']
['Monthly' 'Yearly' 'month' 'annual' 'MONTHLY' 'YEARLY' 'yearly' 'monthly'
 'Annual']
['No' 'Yes' 'no' 'yes' '0' 'Y' 'YES' 'NO' 'n' '1' 'y' 'N']


In [32]:
data['subscription_plan'] = data['subscription_plan'].str.replace('Ent', 'Enterprise').str.strip().str.lower()
data['contract_type'] = data['contract_type'].replace({
                                                        'month': 'Monthly',
                                                        'annual': 'Annual'}).str.strip().str.lower()
data['onboarding_completed'] = data['onboarding_completed'].replace({
                                                                        '0': 'no',
                                                                        'Y': 'yes',
                                                                        'n': 'no',
                                                                        '1': 'yes',
                                                                        'y': 'yes',
                                                                        'N': 'no'}).str.strip().str.lower()

In [33]:
print(data['subscription_plan'].unique())
print(data['contract_type'].unique())
print(data['onboarding_completed'].unique())

['basic' 'enterpriseerprise' 'pro' 'free' 'enterprise']
['monthly' 'yearly' 'annual']
['no' 'yes']


### Handling remanining Nulls.

Numeric columns: fill nulls with median (robust to outliers)
customer_satisfaction_score: 9,385 nulls → fill with median
email_open_rate: 11,710 nulls → fill with median
training_sessions_attended: 10,507 nulls → fill with median
signup_date: 3,521 nulls → drop rows (date is needed for feature engineering)

In [34]:
data.isnull().sum()[data.isnull().sum() > 0]

monthly_revenue                 4692
seats_licensed                  4628
feature_usage_score             5740
last_login_days_ago             6990
integration_count               6964
avg_response_time               8036
customer_satisfaction_score     9238
discount_applied                5770
email_open_rate                11536
training_sessions_attended     10360
signup_date                     3461
dtype: int64

In [35]:
numeric_cols_with_nulls = [
    'monthly_revenue', 'seats_licensed', 'feature_usage_score',
    'last_login_days_ago', 'integration_count', 'avg_response_time',
    'customer_satisfaction_score', 'email_open_rate',
    'training_sessions_attended', 'discount_applied'
]

for col in numeric_cols_with_nulls:
    median_val = data[col].median()
    data[col] = data[col].fillna(median_val)

data = data.dropna(subset=['signup_date'])

print(f"The number of columns with nulls are:- \n{data.isnull().sum()[data.isnull().sum() > 0]}")
print("\nFinal shape:", data.shape)

The number of columns with nulls are:- 
Series([], dtype: int64)

Final shape: (112179, 27)


In [46]:
data.to_parquet('/kaggle/working/cleaned_data.parquet')